core/vm/evmc.go

```diff
@@ -0,0 +1,282 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package vm
+
+import (
+	"math/big"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/ethereum/evmc/bindings/go/evmc"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+type EVMC struct {
+	instance   *evmc.Instance
+	env        *EVM
+	intPool    *intPool
+	readOnly   bool
+	returnData []byte // Last CALL's return data for subsequent reuse
+}
+
+var (
+	createMu     sync.Mutex
+	evmcInstance *evmc.Instance
+)
+
+func createVM(vmPath string) *evmc.Instance {
+	createMu.Lock()
+	defer createMu.Unlock()
+
+	if evmcInstance == nil {
+		if len(vmPath) == 0 {
+			vmPath = os.Getenv("EVMC_PATH")
+			if len(vmPath) == 0 {
+				panic("EVMC vmPath not provided, use --vm flag or set EVMC_PATH")
+			}
+		}
+
+		var err error
+		evmcInstance, err = evmc.Load(vmPath)
+		if err != nil {
+			panic(err.Error())
+		}
+		log.Info("EVMC VM loaded", "name", evmcInstance.Name(), "version", evmcInstance.Version(), "path", vmPath)
+
+		for _, option := range strings.Split(os.Getenv("EVMC_OPTIONS"), " ") {
+			if idx := strings.Index(option, "="); idx >= 0 {
+				name := option[:idx]
+				value := option[idx+1:]
+				err := evmcInstance.SetOption(name, value)
+				if err == nil {
+					log.Info("EVMC VM option set", "name", name, "value", value)
+				} else {
+					log.Warn("EVMC VM option setting failed", "name", name, "error", err)
+				}
+			}
+		}
+	}
+	return evmcInstance
+}
+
+func NewEVMC(env *EVM, cfg Config) *EVMC {
```

Comment:
`EVMC` should implement the `Interpreter` interface, see https://github.com/ethereum/go-ethereum/blob/master/core/vm/interpreter.go#L48